### ISSUE TITLE
Endpoints disappear from ServiceInfo cache

### ISSUE THREAD

**Issue by Sh4d1** [cilium/cilium#11587]
## Bug report

Couldn't find a better title sorry!

**General Information**

- Cilium version (run `cilium version`)
```
Client: 1.7.3 952090308 2020-04-29T15:29:53-07:00 go version go1.13.10 linux/amd64
Daemon: 1.7.3 952090308 2020-04-29T15:29:53-07:00 go version go1.13.10 linux/amd64
```
- Kernel version (run `uname -a`)
```
5.3.0-51-generic #44~18.04.2-Ubuntu SMP Thu Apr 23 14:27:18 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
- Orchestration system version in use (e.g. `kubectl version`, Mesos, ...)
```
Kubernetes
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:48:36Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

Cilium config:
```
  auto-direct-node-routes: "false"
  bpf-ct-global-any-max: "262144"
  bpf-ct-global-tcp-max: "524288"
  cluster-name: default
  cni-chaining-mode: portmap
  container-runtime: none
  debug: "false"
  enable-ipv4: "true"
  enable-ipv6: "false"
  enable-node-port: "false"
  enable-remote-node-identity: "true"
  enable-well-known-identities: "false"
  identity-allocation-mode: crd
  install-iptables-rules: "true"
  kube-proxy-replacement: probe
  masquerade: "true"
  monitor-aggregation: medium
  preallocate-bpf-maps: "false"
  sidecar-istio-proxy-image: cilium/istio_proxy
  tofqdns-enable-poller: "false"
  tunnel: vxlan
  wait-bpf-mount: "false"
```

**How to reproduce the issue**

1. Looks like it happens after some time
2. Run `dig kubernetes.default.svc.cluster.local` inside a pod:
```
;; reply from unexpected source: 100.64.1.197#53, expected 10.32.0.10#53
```
```
$ tcpdump -i any 'port 53'
18:01:58.976876 IP 100.64.1.107.49805 > 10.32.0.10.domain: 49725+ [1au] A? kubernetes.default.svc.cluster.local. (77)
18:01:58.976980 IP 100.64.1.107.49805 > 100.64.1.197.domain: 49725+ [1au] A? kubernetes.default.svc.cluster.local. (77)
18:01:58.977420 IP 100.64.1.197.domain > 100.64.1.107.49805: 49725*- 1/0/1 A 10.32.0.1 (129)
18:01:58.977468 IP 100.64.1.197.domain > 100.64.1.107.49805: 49725*- 1/0/1 A 10.32.0.1 (129)
```
`100.64.1.107` is the pod IP
`10.32.0.10` is coredns service IP
`100.64.1.197` is coredns pod IP

on a working cluster : 
```
18:00:45.324371 IP 100.64.0.86.35473 > 100.64.0.246.domain: 25350+ [1au] A? kubernetes.default.svc.cluster.local. (77)
18:00:45.324444 IP 100.64.0.86.35473 > 100.64.0.246.domain: 25350+ [1au] A? kubernetes.default.svc.cluster.local. (77)
18:00:45.324790 IP 100.64.0.246.domain > 100.64.0.86.35473: 25350*- 1/0/1 A 10.32.0.1 (129)
18:00:45.324813 IP 100.64.0.246.domain > 100.64.0.86.35473: 25350*- 1/0/1 A 10.32.0.1 (129)
```

A restart of the cilium pod solves the issue.

**Comment by Sh4d1** [comment]
On ciliun monitor I also get this extra line:
```
-> stack flow 0xea82d15b identity 40873->2 state new ifindex 0 orig-ip 0.0.0.0: 100.64.1.107:47281 -> 10.32.0.10:53 udp
```

**Comment by Sh4d1** [comment]
Huh, also:
```
$ cilium service list     
ID   Frontend        Service Type   Backend                  
1    10.32.0.1:443   ClusterIP      1 => 51.159.74.75:6443   
```

I don't have the other services

**Comment by brb** [comment]
Could you attach cilium-agent logs from the run when you hit the issue?

**Comment by Sh4d1** [comment]
Looks like the endpoints are missing from the cache :thinking: 
But I can see them in cilium

**Comment by brb** [comment]
What's about `cilium bpf lb list` output?

**Comment by Sh4d1** [comment]
```
SERVICE ADDRESS   BACKEND ADDRESS             
10.32.0.1:443     0.0.0.0:0 (1) [ClusterIP]   
                  51.159.74.75:6443 (1) 
```

**Comment by brb** [comment]
The service entry for kube-dns is missing. Seems that something is wrong with your kube-apiserver(?):
```
level=warning msg="Network status error received, restarting client connections" error="an error on the server (\"[+]ping ok\\n[+]log ok\\n[-]etcd failed: reason withheld\\n[+]poststarthook/start-kube-apiserver-admission-initializer ok\\n[+]poststarthook/generic-apiserver-start-informers ok\\n[+]poststarthook/start-apiextensions-informers ok\\n[+]poststarthook/start-apiextensions-controllers ok\\n[+]poststarthook/crd-informer-synced ok\\n[+]poststarthook/bootstrap-controller ok\\n[+]poststarthook/rbac/bootstrap-roles ok\\n[+]poststarthook/scheduling/bootstrap-system-priority-classes ok\\n[+]poststarthook/start-cluster-authentication-info-controller ok\\n[+]poststarthook/start-kube-aggregator-informers ok\\n[+]poststarthook/apiservice-registration-controller ok\\n[+]poststarthook/apiservice-status-available-controller ok\\n[+]poststarthook/kube-apiserver-autoregistration ok\\n[+]autoregister-completion ok\\n[+]poststarthook/apiservice-openapi-controller ok\\nhealthz check failed\") has prevented the request from succeeding (get healthz.meta.k8s.io)" subsys=k8s
```

**Comment by Sh4d1** [comment]
All service entries except kubernetes are missing 🤔
Regarding the health check yes there is some random failure (but it should not impact Cilium) (and they occur rarely)

**Comment by brb** [comment]
Try to exec to cilium-agent pod, and then do `curl -vvv https://10.32.0.1`.

**Comment by Sh4d1** [comment]
```
curl -k -vvv https://10.32.0.1
* Rebuilt URL to: https://10.32.0.1/
*   Trying 10.32.0.1...
* TCP_NODELAY set
* Connected to 10.32.0.1 (51.159.74.75) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Unknown (8):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Request CERT (13):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Client hello (1):
* TLSv1.3 (OUT), TLS Unknown, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Certificate (11):
* TLSv1.3 (OUT), TLS Unknown, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=kube-apiserver
*  start date: May  6 14:37:06 2020 GMT
*  expire date: May  6 14:37:06 2030 GMT
*  issuer: CN=kubernetes
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* TLSv1.3 (OUT), TLS Unknown, Unknown (23):
* TLSv1.3 (OUT), TLS Unknown, Unknown (23):
* TLSv1.3 (OUT), TLS Unknown, Unknown (23):
* Using Stream ID: 1 (easy handle 0x55b669c8b580)
* TLSv1.3 (OUT), TLS Unknown, Unknown (23):
> GET / HTTP/2
> Host: 10.32.0.1
> User-Agent: curl/7.58.0
> Accept: */*
> 
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS Unknown, Unknown (23):
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
* TLSv1.3 (OUT), TLS Unknown, Unknown (23):
* TLSv1.3 (IN), TLS Unknown, Unknown (23):
* TLSv1.3 (IN), TLS Unknown, Unknown (23):
* TLSv1.3 (IN), TLS Unknown, Unknown (23):
< HTTP/2 403 
< content-type: application/json
< x-content-type-options: nosniff
< content-length: 233
< date: Mon, 18 May 2020 19:15:58 GMT
< 
* TLSv1.3 (IN), TLS Unknown, Unknown (23):
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "forbidden: User \"system:anonymous\" cannot get path \"/\"",
  "reason": "Forbidden",
  "details": {
    
  },
  "code": 403
* Connection #0 to host 10.32.0.1 left intact
}
```

The errors show are just some timeouts we gets from etcd, but they happend rarely.
```
curl -k https://10.32.0.1/healthz
ok
```

**Comment by brb** [comment]
If you restart the cilium-agent, does it fix the issue (i.e. it creates the kube-dns svc)?

**Comment by Sh4d1** [comment]
Yes restarting Cilium fixes the issue

**Comment by Sh4d1** [comment]
@brb though there is no `kube-dns` service, only `coredns`

**Comment by Sh4d1** [comment]
If I delete the pod it also brings back the missing service, but only it's service, not the others.
I noticed all the resync period are set to 0 on the informer, it could be linked

**Comment by brb** [comment]
Can you attach cilium sysdump (if it contains any sensitive information, you can DM (`martynas` on Cilium slack))?

**Comment by Sh4d1** [comment]
Sent via MP. 
Another thing:
The pod that I deleted recreated the endpoint in the ServiceCache, but now it disappeared again

**Comment by aanm** [comment]
@Sh4d1 please add `enable-k8s-endpoint-slice: false` in your config map and try again

**Comment by Sh4d1** [comment]
@aanm I'll see if I can reproduce with `enable-k8s-endpoint-slice: false` , just need to wait to see if it reappears :sweat_smile: 
(though ep slices are enabled on the cluster). Do think it can be caused by this?

**Comment by brb** [comment]
@Sh4d1 Have you reproduce it with the new setting?

**Comment by aanm** [comment]
ping @Sh4d1

**Comment by Sh4d1** [comment]
oops sorry didn't get the previous mention :sweat_smile: 

So I create a bunch of cluster (with and without `enable-k8s-endpoint-slice: false`) and only one hit the issue after 1 or 2 weeks.

In the debug log I can see:
```
2020-06-02T17:10:41.901335025Z level=debug msg="Kubernetes service definition changed" action=service-deleted endpoints= k8sNamespace=kube-system k8sSvcName=coredns old-service=nil service="frontend:10.32.0.10/ports=[dns dns-tcp metrics]/selector=map[k8s-app:kube-dns]" subsys=k8s-watcher
2020-06-02T17:10:41.901405547Z level=debug msg="created new L3n4Addr" ipAddr="{10.32.0.10 {UDP 53}}" subsys=loadbalancer
2020-06-02T17:10:41.901417911Z level=debug msg="created new L3n4Addr" ipAddr="{10.32.0.10 {TCP 9153}}" subsys=loadbalancer
2020-06-02T17:10:41.901938832Z level=debug msg="Deleting service" backends="[{5 scw-dashboard-default-1badc34d8be447beb34cce53 {100.64.0.89 {UDP 53}}}]" serviceID=5 serviceIP="{10.32.0.10 {UDP 53}}" subsys=service
2020-06-02T17:10:41.901991952Z level=debug msg="Deleting obsolete backend" backendID=5 backends="[{5 scw-dashboard-default-1badc34d8be447beb34cce53 {100.64.0.89 {UDP 53}}}]" serviceID=5 serviceIP="{10.32.0.10 {UDP 53}}" subsys=service
2020-06-02T17:10:41.902782992Z level=debug msg="# cilium lb delete-service 10.32.0.10 53 0" k8sNamespace=kube-system k8sSvcName=coredns subsys=k8s-watcher
2020-06-02T17:10:41.902849056Z level=debug msg="Deleting service" backends="[{6 scw-dashboard-default-1badc34d8be447beb34cce53 {100.64.0.89 {TCP 9153}}}]" serviceID=6 serviceIP="{10.32.0.10 {TCP 9153}}" subsys=service
2020-06-02T17:10:41.902864565Z level=debug msg="Deleting obsolete backend" backendID=6 backends="[{6 scw-dashboard-default-1badc34d8be447beb34cce53 {100.64.0.89 {TCP 9153}}}]" serviceID=6 serviceIP="{10.32.0.10 {TCP 9153}}" subsys=service
2020-06-02T17:10:41.902891275Z level=debug msg="# cilium lb delete-service 10.32.0.10 9153 0" k8sNamespace=kube-system k8sSvcName=coredns subsys=k8s-watcher
```

**Comment by brb** [comment]
@Sh4d1 Can you check (maybe with the help of `kubectl -n kube-system get evets`) whether the kube-dns Service object was not removed?

**Comment by brb** [comment]
What's about Endpoint objects?

**Comment by brb** [comment]
Thanks. Can you attach the full cilium-agent log? If it contains sensitive info, then just ~30 lines above the svc deletion.

**Comment by Sh4d1** [comment]
Full log is here https://easyupload.io/fcawuz (1M+ lines with debug :sweat_smile: ) 
`grep "definition change" -C 100` is here : https://gist.github.com/Sh4d1/b5585b1a284f665fd53e7a2ad0cfd593

**Comment by aanm** [comment]
@Sh4d1 This issue is fixed in master and will be part of the 1.7.5 release in the next days. Thanks for reporting the bug.

**Comment by Sh4d1** [comment]
@aanm great thanks :tada: :smile:
